core/src/main/java/org/lflang/federated/generator/FederateInstance.java

```diff
@@ -559,6 +559,76 @@ private void indexExcludedTopLevelReactions(Reactor federatedReactor) {
     }
   }
 
+  /** Cached result for isInZeroDelayCycle(). */
+  private boolean _isInZeroDelayCycle = false;
+
+  /**
+   * Indicator that _isInZeroDelayCycle has been calculated. This will need to be reset if and when
+   * mutations are supported.
+   */
+  private boolean _isInZeroDelayCycleCalculated = false;
+
+  /**
+   * Return true if there is a zero-delay path from this federate to itself. This is used to
+   * determine whether absent messages need to be sent. Note that this is not the same as causality
+   * loop detection. A federate may be in a zero-delay cycle (ZDC) even if there is no causality
+   * loop.
+   */
+  public boolean isInZeroDelayCycle() {
+    if (_isInZeroDelayCycleCalculated) return _isInZeroDelayCycle;
+    _isInZeroDelayCycleCalculated = true;
+    var visited = new HashSet<FederateInstance>();
+    return _isInZeroDelayCycleInternal(this, this, visited);
+  }
+
+  /** Internal helper function for isInZeroDelayCycle(). */
+  private boolean _isInZeroDelayCycleInternal(
+      FederateInstance end, FederateInstance next, HashSet<FederateInstance> visited) {
+    next.sendsTo.forEach(
+        (destination, setOfDelays) -> {
+          // Return if we've already found a cycle.
+          // Also skip self loops because these get optimized away.
+          // Also skip any we've visited.
+          if (end._isInZeroDelayCycle
+              || (end == next && destination == next)
+              || visited.contains(destination)) return;
+          visited.add(destination);
+          if (setOfDelays.contains(null)) {
+            // There is a zero-delay connection to destination.
+            if (destination == end) {
+              // Found a zero delay cycle.
+              end._isInZeroDelayCycle = true;
+              return;
+            }
+            _isInZeroDelayCycleInternal(end, destination, visited);
+          }
+        });
+    return end._isInZeroDelayCycle;
+  }
+
+  /**
+   * Return true if there is a downstream federate that is in a zero-delay cycle. This is needed
+   * because such a federate may be called upon to generate absent messages even if it is not itself
+   * in a zero-delay cycle. An example is the test `test/C/src/federated/FeedbackDelay.lf`. The
+   * absent message may be needed even if there is a delay on the connection to the downstream
+   * federate.
+   */
+  public boolean isUpstreamOfZeroDelayCycle() {
```

Comment:
I don't understand why this check is needed. Because of the below code, the network message action of `reaction_1` of the federate c in `FeedbackDelay.lf` is not in `zeroDelayCycleNetworkMessageActionTable`. So, the federate c doesn't wait for port absent messages from the federate pl. Then, why does the federate pl send ABS messages?
https://github.com/lf-lang/lingua-franca/blob/279b3fd1c2b7f5e7b365e759e60785611ff60d18/core/src/main/java/org/lflang/federated/generator/FedASTUtils.java#L280

![image](https://github.com/lf-lang/lingua-franca/assets/78055940/c9851a79-d0b2-4aa2-81c2-140d28f5b746)

Reply:
Good catch again!!  Fixed in 224356488cef0b60dab569b4c24f7213ea86449e